bigframes/core/compile/polars/compiler.py

```diff
@@ -0,0 +1,373 @@
+                return expr
+            raise NotImplementedError(f"Polars compiler hasn't implemented {op}")
+
+    @dataclasses.dataclass(frozen=True)
+    class PolarsAggregateCompiler:
+        scalar_compiler = PolarsExpressionCompiler()
+
+        def get_args(
+            self,
+            agg: ex.Aggregation,
+        ) -> Sequence[pl.Expr]:
+            if isinstance(agg, ex.NullaryAggregation):
+                return []
+            elif isinstance(agg, ex.UnaryAggregation):
+                arg = self.scalar_compiler.compile_expression(agg.arg)
+                return [arg]
+            elif isinstance(agg, ex.BinaryAggregation):
+                larg = self.scalar_compiler.compile_expression(agg.left)
+                rarg = self.scalar_compiler.compile_expression(agg.right)
+                return [larg, rarg]
+
+            raise NotImplementedError(
+                f"Aggregation {agg} not yet supported in polars engine."
+            )
+
+        def compile_agg_op(self, op: agg_ops.WindowOp, inputs: Sequence[str] = []):
+            if isinstance(op, agg_ops.ProductOp):
+                # TODO: Need schema to cast back to original type if posisble (eg float back to int)
+                return pl.col(*inputs).log().sum().exp()
+            if isinstance(op, agg_ops.SumOp):
+                return pl.sum(*inputs)
+            if isinstance(op, agg_ops.MinOp):
+                return pl.min(*inputs)
+            if isinstance(op, agg_ops.MaxOp):
+                return pl.max(*inputs)
+            if isinstance(op, agg_ops.CountOp):
+                return pl.count(*inputs)
+            if isinstance(op, agg_ops.CorrOp):
+                return pl.corr(*inputs)
+            raise NotImplementedError(
+                f"Aggregate op {op} not yet supported in polars engine."
+            )
+
+
+@dataclasses.dataclass(frozen=True)
+class PolarsCompiler:
+    """
+    Compiles ArrayValue to polars LazyFrame and executes.
+
+    This feature is in development and is incomplete.
+    While most node types are supported, this has the following limitations:
+    1. GBQ data sources not supported.
+    2. Joins do not order rows correctly
+    3. Incomplete scalar op support
+    4. Incomplete aggregate op support
+    5. Incomplete analytic op support
+    6. Some complex windowing types not supported (eg. groupby + rolling)
+    7. UDFs are not supported.
+    8. Returned types may not be entirely consistent with BigQuery backend
+    9. Some operations are not entirely lazy - sampling and somse windowing.
+    """
+
+    expr_compiler = PolarsExpressionCompiler()
+    agg_compiler = PolarsAggregateCompiler()
+
+    def compile(self, array_value: bigframes.core.ArrayValue) -> pl.LazyFrame:
+        if not polars_installed:
+            raise ValueError(
+                "Polars is not installed, cannot compile to polars engine."
+            )
+
+        # Polars has incomplete slice support in lazy mode
```

Comment:
Maybe TODO to generalize BFET -> BFET rewrites

Reply:
added comment